src/components/app.tsx

```diff
@@ -1,16 +1,176 @@
+        audioSource.current?.start();
+
+      const startTime = audioContext.current.currentTime;
+
+      const measureProgress = () => {
+        if (playingRef.current && audioContext.current && audioBuffer.current) {
+          setPlaybackProgress(playbackRate * (audioContext.current.currentTime - startTime) / audioBuffer.current.duration);
+          requestAnimationFrame(measureProgress);
+        }
+      };
+      measureProgress();
+    } else {
+      audioSource.current?.stop();
+    }
+  }, [playing, playbackRate]);
+
+  const handleSoundChange = (event: ChangeEvent<HTMLSelectElement>) => {
+    setSelectedSound(event.currentTarget.value as SoundName);
+  };
+
+  const handleVolumeChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setVolume(Number(event.target.value));
+  };
+
+  const handlePlay = () => {
+    setPlaying(!playing);
+  };
+
+  const handleZoomIn = () => {
+    setZoom(Math.min(512, zoom * 2));
+  };
+
+  const handleZoomOut = () => {
+    setZoom(Math.max(2, zoom * 0.5));
+  };
+
+
+  const handleFasterPlayback = () => {
+    setPlaybackRate(Math.min(8, playbackRate * 2));
+  };
+
+  const handleSlowerPlayback = () => {
+    setPlaybackRate(Math.max(1 / 128, playbackRate * 0.5));
+  };
+
   return (
     <div className="app">
-      <img src={Icon}/>
-      <Text text={sampleText} />
+      <div className="header"><BSCSLogo /> Sounds are waves</div>
+      <div>
+        <select className="sound-picker" value={selectedSound} onChange={handleSoundChange}>
+          <option value="middle-c">Middle C (261.65Hz)</option>
+          <option value="c2">A (65.41 Hz)</option>
+          <option value="baby-cry">Baby Cry</option>
+        </select>
+      </div>
+      <div className="playback">
+        <button onClick={handlePlay}>{ playing ? "Pause" : "Play" }</button>
+        <input type="range" min="0" max="2" step="0.01" value={volume} onChange={handleVolumeChange} />
+      </div>
+      <div className="sound-wave-container">
+        <SoundWave
+          width={600} height={200}
+          data={soundWaveData}
+          volume={volume}
+          playbackProgress={playbackProgress}
+          drawingStep={Math.max(1, 128 / zoom)} // draw every Nth data point, based on the current zoom level
+          zoom={zoom}
+          zoomedInView={true}
+        />
+        <SoundWave
+          width={600} height={60}
+          data={soundWaveData}
+          volume={volume}
+          playbackProgress={playbackProgress}
+          drawingStep={64} // draw every 64th data point
+          zoom={zoom}
+          zoomedInView={false}
+        />
+      </div>
+      <div className="zoom-controls">
+        Zoom: { zoom }
+        <button onClick={handleZoomIn}>Zoom In</button>
+        <button onClick={handleZoomOut}>Zoom Out</button>
+      </div>
+      <div className="zoom-controls">
+        Playback Rate: { playbackRate >= 1 ?  playbackRate : `1/${Math.round(1/playbackRate)}` }x
```

Comment:
```suggestion
        Playback Rate: { playbackRate >= 1 ? playbackRate : `1/${Math.round(1/playbackRate)}` }x
```
Just to prove I'm paying attention. 😉